### ISSUE TITLE
RFC: Plugins monorepo

### ISSUE THREAD

**Issue by Rich-Harris** [ractivejs/ractive#2271]
Plugins are something of a ghetto at the moment. Clicking [Plugins](http://docs.ractivejs.org/latest/plugins) in the nav of any ractivejs.org page takes you to a page on docs.ractivejs.org (with the Docs tab highlighted), containing a bunch of links, many of which are to out-of-date pages (I've been pretty dreadful at keeping the 'official' ones up-to-date).

On top of that, the process for creating plugins is unnecessarily confusing and inconsistent and out-of-date (hello, unmaintained Grunt template).

## Solution: a monorepo

Specifically, [this monorepo](https://github.com/ractivejs/ractive-plugins). There are a few clear advantages:

* We can automate lots of drudge work – building dist files, generating high-quality package.json files, etc
* Much easier to maintain consistency of code style and so forth
* Creating new plugins is as simple as creating a new folder and cribbing from similar plugins. No awkward project setup. Everyone gets ES6 out of the box
* Discoverability is much better – for plugin users, plugin developers (easier to avoid duplication of effort), Ractive developers (at the moment it's hard to keep track of which plugins are in the wild)
* When plugin APIs change (which is inevitable, e.g. we'll probably need to overhaul the adaptor and transition APIs at some point) it's much easier to update plugins en masse
* Easier to have a coherent testing strategy
* It'll make it much easier to generate a plugins.ractivejs.org site, with (for example) automatically generated installation instructions and demos for each plugin

## Downsides

* Some loss of flexibility – if a plugin needed a very specific build process, we'd have to accommodate that somehow
* Third party plugin authors would need to dance to our tune if they wanted to be in the official plugin repo (maybe not a bad thing though – easier to ensure quality across the ecosystem)
* The repo *could* get huge, eventually

Lots still to do but I thought I'd put this out there to gauge initial reactions. Let me know your thoughts

**Comment by JonDum** [comment]
How is it going to work with NPM? How will you make tags and versions if it's all a single repo?

How will we standardize on templates & css? Or is this *only* for transitions, adaptors, events. What about decorators and components then? Rollup doesn't handle anything besides js, right? My "solution" to this has been webpack with the side benefit of being able to point the package.main to the `src/` file and completely avoid the additional umd/loader bloat (but you'd have to be using webpack also :|).  

What about issues? Are we ok with having a clusterfuck of issues all in one repo? I can already guarantee some people won't even mention which plugin they're talking about when creating an issue or asking for help.

I don't know how it's worked for Babel, but as a counter example Grunt tasks tend to have fairly a consistent ecosystem due to [grunt-init-plugin](https://github.com/gruntjs/grunt-init-gruntplugin) generator. No one cares to stray too far from it since it does 95% of the work for you with one command.

**Comment by Rich-Harris** [comment]
@JonDum Lots of good questions.

> How is it going to work with NPM? How will you make tags and versions if it's all a single repo?

Things wouldn't be much different with npm:

```bash
node scripts/build.js # or whatever this ends up being
cd plugins/ractive-foo-bar
npm publish
```

Ideally, plugin authors would `npm add owner ractive` (that user doesn't currently exist, but probably should) so that the core team would be able to help keep packages updated.

Plugins would still be versioned individually.

Tags are a little trickier, but even here I think this approach ultimately makes life easier than what we have right now. As far as I can tell, the main reason people care about tags is so that they can get dist files for a specific commit. Frankly, people using GitHub as a package manager (i.e. Bower users) are doing it wrong – GitHub isn't a place to store dist files. But with a monorepo, we could easily have a bunch of 'dummy' repos, registered with Bower, that *do* store dist files:

```bash
# this could
#  1. build the plugin (bundle, and update package.json from plugin.json)
#  2. run the tests
#  3. publish to npm
#  4. copy dist files to git://github.com/JonDum/ractive-foo-bar
#  5. add tags to git://github.com/JonDum/ractive-foo-bar
node scripts/release.js plugins/ractive-foo-bar
```

> How will we standardize on templates & css? Or is this only for transitions, adaptors, events. What about decorators and components then? Rollup doesn't handle anything besides js, right?

I think it's for everything. We haven't really dealt with components at an ecosystem level – it's a little ad hoc at the moment. I think component packages should contain a) the component file itself (vanilla JS and CSS only, though the source could conceivably use preprocessors and/or linked files), and b) JavaScript versions thereof, so that both of these work:

```html
<link rel='ractive' href='ractive-components-widget/Widget.html'>
<Widget/>
```

```js
import Widget from 'ractive-components-widget';
new Widget({ el: 'main' });
```

I'm very keen to avoid situations where plugin consumers need to use specific tooling (whether that means Rollup or Webpack or a specific component loader or whatever).

> What about issues? Are we ok with having a clusterfuck of issues all in one repo? I can already guarantee some people won't even mention which plugin they're talking about

I'd rather have a clusterfuck of issues in one repo than a clusterfuck of issues across a clusterfuck of repos :grinning: Admins would have the ability to edit issue titles (i.e. 'it doesnt work lol' becomes '[ractive-transitions-slide] it doesnt work lol'), so it shouldn't be too tricky to filter and manage. Issues that came up with multiple plugins would be easier to identify and deal with efficiently.

> Grunt tasks tend to have fairly a consistent ecosystem due to grunt-init-plugin generator

The problem isn't creating plugins so much as maintaining them, and their myriad devDependencies. I'm old enough to remember the chaos that resulted when Grunt 0.4 came out with a different API from 0.3... Also, I just created a Grunt plugin and it was 38Mb! Sharing the bulky devDependencies between plugins results in a much nicer time, I believe.

**Comment by evs-chris** [comment]
I think I'm +1 on the ractive-plugins repo. I don't know about pulling in third parties, but having all of the official plugins (src) in one place sounds like a good thing. It would certainly make for a better how-to-build-plugins reference. FWIW I tend not to try to fix plugin issues because a) I've never worked with/on one before, b) they all appear to be a little different, and c) I'm too lazy/have too much going on to resolve a and b. I think a monorepo would solve that.

Slight tangent, I think the array and magic wrappers should be among the first new residents in the new repo after the existing community is brought together. Having automated testing against ractive from ractive-plugins would probably keep that from getting too hairy with breakage. But after 0.8 :smile:

**Comment by matthew-dean** [comment]
As a consumer, +1 to one repo. Because the issues for an individual are so light, and the documentation doesn't need much out of necessity, making one plugin per repo makes the repo seem like abandonware. That has been my impression of plugins thus far; that is, that they're somewhat of an afterthought and not really core and no one on the Ractive team actually expects you to use them, and if you do, they likely won't work. 

That's not to say any of those are true, but a single ractive-plugins repo makes it seem like plugins are a first-class citizen of the library.

**Comment by JonDum** [comment]
> As far as I can tell, the main reason people care about tags is so that they can get dist files for a specific commit...Frankly, people using GitHub as a package manager (i.e. Bower users) are doing it wrong – GitHub isn't a place to store dist files.

Yes and no. I'd say it's more so that you know with relative confidence which commit is the one published to npm for that version. That's why running `npm version minor` automatically bumps the `version` in the `package.json` **and** makes a new commit for you with that change and a subsequent tag of the new version (so unless you're rewriting history with `rebase -i`, the developer can't mess this up). Say I want to checkout an old version of something as it was at `v0.2.1` on npm to apply a patch and test something—I would just need to run `git checkout tags/v0.2.1`. With everything in one repo, we definitely lose that without a bunch of workarounds.

> I think component packages should contain a) the component file itself (vanilla JS and CSS only, though the source could conceivably use preprocessors and/or linked files),

And the template(s)? For those rare few of us who use runtime.js.

> I'd rather have a clusterfuck of issues in one repo than a clusterfuck of issues across a clusterfuck of repos

Sounds good for core team. What's the work flow for us non-core people? Hundreds of PRs?

If it's easier for you guys to have the plugins *you* maintain all in a single repo, then by all means. No complaints about that, should definitely go ahead and merge them all in. I'm not sure I'm comfortable with the workflow/tools yet for the stuff that my team uses across a bunch of projects.

---- 

PS

> Also, I just created a Grunt plugin and it was 38Mb! Sharing the bulky devDependencies between plugins results in a much nicer time, I believe.

Add `link=true` to your `~/.npmrc`. Then everything is symlinked :)

**Comment by rstacruz** [comment]
the only strange thing here is that you can't tag releases properly. In the rails/rails monorepo, this is fine because all the sub-packages have synchronized version numbers. that's not the case with plugins.

**Comment by fskreuz** [comment]
So this is similar to the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) repo containing TypeScript interfaces for a bazillion different libraries? Or [cdnjs's repo](https://github.com/cdnjs/cdnjs/tree/master/ajax/libs) containing a bazillion libs? I think I'll pass.

How about a monorepo that just "links" to other repos? Kinda like git submodules or a package.json. This makes the monorepo just a list and not the actual repo containing the code. First thing that comes into mind is something like the [microjs.com's repo](https://github.com/madrobby/microjs.com) which is just a list of links, and not the repo containing all the tiny JS libraries that are advertised on that site.

Plugins that want to be official plugins should be in a repo that follow a certain convention in the way they're written, how the files are structured, what interfaces should be exposed, what branches should exist in the repo. That way, a centrally controlled tool can easily discover the files, run the test against the plugin, easily build files off of them, etc.

To make it easy for other devs, they could fork from a "ractive plugin starter" repo or some Yeoman scaffold containing the folder structure, linting configuration, testing configurations etc. This scaffold/starter repo can be maintained by the core team along with the list repo.

Then when all of that is done, all the third-party dev has to do is submit a PR to the list repo which adds their plugin link. All the core team needs to do is inspect his fork and when all is good, merges it in. Repos that don't follow convention will just fall off the official status unless the author has concerns.

**Comment by matthew-dean** [comment]
@fskreuz I'm not sure how DefinitelyTyped is a good example, when it seems logical for them to be in one repo. I'm not sure how one would find TS definition files for X library without it.

> Managing a thousand things (and forking a thousand things just to add one thing) looks like a terrible idea.

You wouldn't be forking a thousand things. You'd be forking the plugin repo. You can only call it a thousand things if you call any repo with a thousand files a thousand things. Many plugins, like transition plugins, are illogical if separate. Maybe the Ractive plugins could be grouped by type? (Types listed here: http://docs.ractivejs.org/latest/plugins .) And beyond that, plugins should be included in one repo only if they have the same maintainer. Otherwise issues wouldn't really be going to the right person.

However, it's done, I think it's also worth keeping a central list of all Ractive plugins anywhere, a la PostCSS (scroll to plugin lists): https://github.com/postcss/postcss

**Comment by fskreuz** [comment]
On the other hand, a central repo sounds good (I'm really split here :\). In Drupal, they have this mantra of "Don't duplicate, but cooperate." (I forgot the exact words, but whatever) where they encourage contributors to not duplicate the effort and instead, contribute on what's existing. A central repo appears to promote this.

Hopefully we don't end up making our own package manager though. We'd end up like Meteor... or Drupal (which I heard is moving over to composer for managing dependencies instead of their drush tool).

2c. 4c in total now.

**Comment by JonDum** [comment]
> A central repo appears to promote this.

I don't know if that's true. It's really about maintenance. Making it easier to contribute promotes contributing.

Small list that I can think of: 

* Discovering existing plugins ( dynamic website based off repos / npm tags?)
* Maintainers quicker to accept PRs / keeping NPM up-
* Standardized stack / build tools / **npm tags** (i.e., `ractive`, `ractive-plugin`, `ractive-transition`, etc) 
* Making a generator/template for new plugins
* etc?

Those things aren't exclusive of a single repo or multiple repos imo.

**Comment by guilhermeaiolfi** [comment]
> Specifically, this monorepo. There are a few clear advantages:
> 
>We can automate lots of drudge work – building dist files, generating high-quality package.json files, etc

That could be done with tools and give more flexibility as needed. A good example here is ember: [Developing Addons and Blueprints](http://www.ember-cli.com/extending/)

>Much easier to maintain consistency of code style and so forth
>Creating new plugins is as simple as creating a new folder and cribbing from similar plugins. No awkward project setup. Everyone gets ES6 out of the box

Tools can enforce the first and do a much better job at the second item.

>Discoverability is much better – for plugin users, plugin developers (easier to avoid duplication of effort), Ractive developers (at the moment it's hard to keep track of which plugins are in the wild)

That's not the git repo responsibility. That could be handled by a list of links (could be a git repo), a different DB, a simple website, etc.

>When plugin APIs change (which is inevitable, e.g. we'll probably need to overhaul the adaptor and transition APIs at some point) it's much easier to update plugins en masse

Well, if you are the author of those plugins, yes. But if the plugins are from different authors, you are touching someone else work. Ultimately, 

>Easier to have a coherent testing strategy
>It'll make it much easier to generate a plugins.ractivejs.org site, with (for example) automatically generated installation instructions and demos for each plugin

Again, something like ember is doing seems great and much better at this.

I'm strongly against this. That doesn't lower the enter barrier to develop plugins, doesn't scale for more than a dozen plugins, you can't manage permissions for different plugins, take out the flexibility that some plugins may need. For plugin authors, needing to PR every change for its own code is a downside.

I prefer a bazillion repos grouped together by a plugin center to made them public, and let the developers decide which one is the best than one big repo with a fewer, more restricted, limited plugins. Even tho those may be more "unified" somehow. 

That seems like what Linux packages tries to do vs what Android/iOS does.

**Comment by dagnelies** [comment]
:-1: 

TLDR;

...seems like a disadvantage to me. It's way easier for me to manage my own little plugins. The big parts of it are developping it, documenting it, maintaining it. And this is definitely easier if you have your own space and dedicated project.

I don't want to:

* checkout huge codebase to work on a little plugin :(
* follow tickets in a "foreign" mono-repo :(
* fix my builds because something changed :(
* fight the pages because I need something else :(

---

What is the problem with the repos anyway?

* plugins are not up-to-date? A mono-repo won't change that. People do.
* how to make a plugin is messy? A mono-repo won't change that. Simplifying the plugin template would.

---

The current shortcoming IMHO, is the following page:

http://docs.ractivejs.org/latest/plugins

It's simply cluttered and unreadable. Perhaps simply working on THAT page would greatly improve it.

1. Take the whole page. Not merely like a doc page.
2. Sort the plugins by "popularity" (or what you think are more in demand)
3. Present them better. For instance something like:

**Plugin Name** *Plugin Type*
(in small: author / repo)
A small Iframe with a demo

**Comment by Madgvox** [comment]
Seems like this comes back to revamping the docs. It's becoming more and more apparent that the docs / website in general should be the No.1 priority after 0.8.
